Step2

oyj7677 opened the pull request:
step1의 개선 사항을 포함하고 step2 진행하여 PR합니다.

Commit:
step1 개선 사항
 - onCreate()내의 로직 중 CounterViewModel의 observe 설정 로직 분리
 - CounterViewModel의 Counter값 String -> Int로 변경, xml에서 데이터 바인딩 시 toString()사용
 - xmlns 최상위 태그로 이동
 - CounterViewModel생성 로직 테스트 삭제
	- 각 테스트 마다 Counter의 초기값이 다를 수 있기 때문에 전역변서 선언 X
 - CounterViewModel의 Counter초기값 파라미터 추가

Commit:
step2 - MVVM변환
 - MVP패턴에 사용된 Contract, Presenter 삭제
 - MVVM패턴을 위한 DataBinding 및 ViewModel생성
 - Toast출력 이벤트를 위한 SingleLiveEvent클래스 추가
 - CalculatorViewModel테스트 코드 작성

wisemuji commented on calculator-app/src/main/java/camp/nextstep/edu/calculator/CalculatorActivity.kt:
```diff
@@ -2,48 +2,30 @@ package camp.nextstep.edu.calculator
-        binding.textView.text = expression.toString()
-    }
+        binding.lifecycleOwner = this
+        binding.viewModel = viewModel
```
다음 코멘트에 대해 고민해보면 어떨까요?
https://github.com/next-step/android-calculator-mvvm/pull/157#discussion_r1293488940

wisemuji commented on calculator-app/src/main/java/camp/nextstep/edu/calculator/CalculatorViewModel.kt:
```diff
@@ -0,0 +1,55 @@
+    fun calculate() {
+        val result = calculator.calculate(expression.toString())
+        if (result == null) {
+            _toastEvent.value = "완성되지 않은 수식입니다"
```
만약 토스트 메시지 요구사항이 바뀌어 해당 문구가 다른 문구로 교체된다면 어떻게 될까요?
교체를 위해 ViewModel을 수정해야 할거고, 결국 View 요구사항 변화로 인해 ViewModel이 바뀌게 되는 것입니다.
MVVM 구조에서는 ViewModel에서 최대한 Android, View 관련 로직을 제거하도록 가이드합니다.

> MVVM의 핵심은 View와 ViewModel 연결을 최소화한다는 점에 있습니다.

wisemuji replied:
또한, 해당 문구는 strings.xml에서 관리해보면 어떨까요?

wisemuji commented on calculator-app/src/main/java/camp/nextstep/edu/calculator/CalculatorViewModel.kt:
```diff
@@ -0,0 +1,55 @@
+
+    init {
+        setFormula()
+    }
```
```kotlin
private var _formula = MutableLiveData<String>(expression.toString())
val formula: LiveData<String>
    get() = _formula
```
위와 같이 작성하면 불필요한 init block을 제거할 수 있습니다 🙂 

아래는 참고만 해주세요!
저는 개인적으로 ViewModel의 init block 내에 로직을 최소화하려고 합니다. 개발자 입장에서는 ViewModel 인스턴스만 만들었을 뿐인데 불필요하게 부수 효과(side effect)가 발생하는 것이 예상치 못한 동작일 수 있기 때문입니다.

wisemuji commented on calculator-app/src/main/java/camp/nextstep/edu/calculator/CalculatorViewModel.kt:
```diff
@@ -0,0 +1,55 @@
+    fun removeLast() {
+        expression = expression.removeLast()
+        setFormula()
+    }
```
매번 변경사항이 생길때마다 setFormula()를 호출해서 알려주기보다, LiveData의 장점을 살려서 바로 변경사항을 반영하면 어떨까요?
힌트: expression과 formula가 굳이 분리될 필요가 있을까요?

wisemuji commented on calculator-app/src/test/kotlin/camp/nextstep/edu/calculator/CalculatorViewModelTest.kt:
```diff
@@ -0,0 +1,126 @@
+    val instantExecutorRule = InstantTaskExecutorRule()
+
+    @Test
+    fun `입력된 숫자가 없을 때, 숫자 클릭`() {
```
테스트명을 통해 **어떤 것을 테스트하고자 하는지** 제대로 드러나지 않는다고 생각합니다. 테스트명을 작성하실 때에는
1. 검증하고자 하는 로직을 명확하게 정하고 (검증이 필요한 로직이 두 개 이상이면 필요에 따라 또 다른 함수로 분할한다)
2. 그것을 함수명으로 드러내보면 어떨까요? 함수명에 기대되는 결과를 같이 드러낸다면 더욱 의미를 파악하기 쉬워집니다.

이하 동일합니다!

wisemuji reviewed (changes requested):
안녕하세요 영재님!

Step2 구현을 잘 해주셨습니다 💯  깔끔한 구현이 좋았습니다
고민해보시면 좋을만한 코멘트를 달았으니 확인해주시면 감사하겠습니다.

궁금한 점이나 논의하고 싶은 내용이 있으면 언제든 DM 주세요! 🙏 💪

Commit:
step2 1차 개선사항
 - onCreate내의 로직들 의미있는 함수들로 분리
	- 추가적으로 Binding관련 로직 분리
 - View요구사항에 따른 ViewModel변경 방지
 	- 토스트 메시지 string.xml관리
	- View에서 string res Id를 받아 getString하여 사용
 - Expression()을 MutableLiveData의 파라미터로 사용하여 기존 Formula역할 흡수
 - ViewModel의 init 삭제
 - 테스트 코드 함수명 변경
	- 함수명에 기대되는 결과 추가

Commit:
Counter 앱 추가 수정
- onCreate내의 로직들 의미있는 함수들로 분리
	- 추가적으로 Binding관련 로직 분리

wisemuji commented on calculator-app/src/main/java/camp/nextstep/edu/calculator/CalculatorViewModel.kt:
```diff
@@ -0,0 +1,43 @@
+import camp.nextstep.edu.calculator.domain.Operator
+import kotlin.coroutines.coroutineContext
+
+class CalculatorViewModel(initFormula: List<Any> = emptyList()) : ViewModel() {
+    private val calculator = Calculator()
+    private var expression = Expression(initFormula)
+
+    private var _formula = MutableLiveData(expression)
+    val formula: LiveData<Expression>
+        get() = _formula
```
initFormula값을 바로 _formula 초기화에 사용해도 되지 않을까요? 지금은 작은 ViewModel이지만 나중에 앱이 커질수록 클래스 내 프로퍼티가 많이 생기게 되어, 이런 값들은 꼭 선언하지 않는 것이 관리하기 편할 수 있어요.
```suggestion
class CalculatorViewModel(initFormula: List<Any> = emptyList()) : ViewModel() {
    private val calculator = Calculator()

    private var _formula = MutableLiveData(Expression(initFormula))
    val formula: LiveData<Expression>
        get() = _formula
```

wisemuji reviewed (approved):
안녕하세요 영재님! 
리뷰 반영을 너무나도 잘 해주셨습니다 💯 

다음 미션도 화이팅입니다 🙏
